### PR TITLE
[#525] Remove and Invalidate API tokens

### LINKED ISSUE #525: Need an option to remove and invalidate API token
Currently EDI server generates JWT tokens based on Secret and Expiration date. Tokens are valid till expiration date. Need to implement a feature to invalidate existing token.

One of the possible solution is:

    Store blacklisted tokens flat in the ConfigMap (one per line)
    Mount blacklisted tokens on EDGEs
    Read blacklisted tokens on start and keep in in memory using Lua
    Monitor them using legion-template tool and reload nginx on each config-map update (as we doing now for statistics)
    On file update blacklist will be reloaded in-memory.
    Request filtering will be done by lua scripts (checking token existence in in-memory array)

### PR DESCRIPTION
**Pull request by leshchanka-aliaksandr**
This closes #525

### CODE REVIEW

**Comment by kirillmakhonin** [comment]
Maybe it would be better to use plain files for reading data? Not K8S API? @aliaksandr-d @dsuslov  @akharlamov ?

File: legion/legion/template.py [hunk 1]
```diff
@@ -122,9 +119,8 @@ def load_module(self, name, *args, **kwargs):
         module_name, function_name = name.rsplit('.', 1)
         module = importlib.import_module(module_name)
         module_function = getattr(module, function_name)
-        future = asyncio.ensure_future(module_function(self, *args, **kwargs))
-
-        self._coroutines.append(future)
+        kwargs['template_system'] = self
+        self._plugins.append((module_function, args, kwargs, ))
```
Comment (kirillmakhonin) [inline legion/legion/template.py]: Redundant

File: k8s/edge/nginx.conf.ltmpl [hunk 2]
```diff
@@ -14,6 +14,7 @@
 #    limitations under the License.
 #
 {{ load_module('legion.template_plugins.enclave_models_monitor') }}
+{{ load_module('legion.template_plugins.enclave_configmap_monitor', key="invalid-tokens") }}
```
Comment (kirillmakhonin) [inline k8s/edge/nginx.conf.ltmpl]: Please use single quote

Comment (kirillmakhonin) [reply k8s/edge/nginx.conf.ltmpl]: Please use specific variable name, not `var`

File: legion/legion/template.py [hunk 3]
```diff
@@ -178,18 +175,19 @@ def render_loop(self):
...
-            task = result[0].pop()
-            if type(task) == Task and task.exception() is not None:
-                raise task.exception()
+        running_threads = []
+        for f, args, kwargs in self._plugins:
+            thread = threading.Thread(target=f, args=args, kwargs=kwargs,
+                                      daemon=True, name=f.__name__)
+            thread.start()
+            running_threads.append(thread)
+        for t in running_threads:
```
Comment (kirillmakhonin) [inline legion/legion/template.py]: Please analyze is it the best solution or not

File: legion/legion/template_plugins/io.py [hunk 4]
```diff
@@ -70,12 +71,30 @@
...
+    event_loop = asyncio.new_event_loop()
+    asyncio.set_event_loop(event_loop)
+    future = asyncio.ensure_future(loop(event_loop, watcher, template_system, filepath, is_yaml, var_name))
+    event_loop.run_until_complete(future)
 
+
+async def loop(event_loop, watcher, template_system, filepath, is_yaml=False, var_name='file'):
+    """
+    Run aionotify watcher in current event loop
+    :param event_loop: Current event loop
```
Comment (kirillmakhonin) [inline legion/legion/template_plugins/io.py]: Pls add space before args

Comment (leshchanka-aliaksandr) [reply legion/legion/template_plugins/io.py]: The problem was resolved by this commit fc7ff0219e338fd298700b3e94636710810690c6

**Comment by kirillmakhonin** [comment]
bot, test app

**Comment by aliaksandr-d** [comment]
bot, test all

File: deploy/helms/legion/templates/edge/edge-rbac.yaml [hunk 5]
```diff
@@ -7,6 +7,13 @@ rules:
 - apiGroups: [""] # core API group
   resources: ["services"]
   verbs: ["watch", "list"]
+- apiGroups: [""] # core API group
+  resources: ["configmaps"]
+  verbs: ["create", "get", "patch", "update"]
+  resourceNames: ["{{ .Release.Name }}-invalid-tokens"]
```
Comment (dsuslov) [inline deploy/helms/legion/templates/edge/edge-rbac.yaml]: @aliaksandr-d  [minor] invalid-tokens ->blacklisted-tokens for all places in this PR where "invalid-tokens" is used

File: k8s/edge/invalid_tokens.lua [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+--   limitations under the License.
+--
+local _M = {}
+
+function _M.check_token()
+    local auth_header = ngx.var.http_Authorization
+
+    local _, _, token = string.find(auth_header, "Bearer%s+(.+)")
+    if _M.invalid_tokens[token] ~= nil then
+        ngx.log(ngx.WARN, "Invalid token")
```
Comment (dsuslov) [inline k8s/edge/invalid_tokens.lua]: token is blacklisted

File: legion/legion/edi/server.py [hunk 7]
```diff
@@ -318,7 +319,8 @@ def generate_token():
         jwt_life_length = timedelta(minutes=int(app.config['JWT_CONFIG']['jwt.length.minutes']))
```
Comment (dsuslov) [inline legion/legion/edi/server.py]: @aliaksandr-d  is token TTL configured globally for all tokens? It should be defined individually of each token

**Comment by aliaksandr-d** [comment]
We decided to improved API Tokens management as a part of  #573

**Comment by aliaksandr-d** [comment]
bot, test all

**Comment by aliaksandr-d** [comment]
bot, test all

**Comment by aliaksandr-d** [comment]
bot, test all

**Pull request merged**